flash/core/serve/dag/optimization.py

```diff
@@ -877,7 +877,7 @@ def __eq__(self, other):
         return is_eq
 
     def __ne__(self, other):
-        return not (self == other)
+        return self != other
```

Comment:
Have to avoid recursion on `__ne__`:
```suggestion
        return not self.__eq__(other)
```